mlir/lib/Conversion/AIRToAIEPass.cpp

```diff
@@ -78,6 +92,41 @@ AIE::TileOp getPhysTileOp(ModuleOp aie_module, int col, int row) {
                                      col, row);
 }
 
+struct ShimTileAllocator {
+
+  std::vector<int> shim_columns;
+  int shim_dma_channels;
+
+  struct shim_allocation_info_t {
+    AIE::TileOp shim_tile;
+    int available_channels;
+  };
+
+  std::vector<shim_allocation_info_t> mm2s_allocs, s2mm_allocs;
+
+  ShimTileAllocator(std::vector<int> cols, int channels)
+      : shim_columns(cols), shim_dma_channels(channels) {}
+
+  AIE::TileOp getShimTile(ModuleOp aie_module, int src_memory_space,
+                          int dst_memory_space) {
+    bool isMM2S = (src_memory_space < dst_memory_space);
+    auto allocs = isMM2S ? &mm2s_allocs : &s2mm_allocs;
+
+    // return first available shim tile with a free channel
```

Comment:
I know this is what the current allocator does, but it's not safe in general to just return the next channel.